deploy/buildtools/buildfarm.py

```diff
@@ -0,0 +1,334 @@
+        dest_build_dir: Name of build dir on build host.
+        ip_address: IP address of build host.
+    """
+    build_config: BuildConfig
+    dest_build_dir: str
+    ip_address: str
+
+    def __init__(self):
+        self.ip_address = ""
+        self.build_config = None # type: ignore
+        self.dest_build_dir = ""
+
+class BuildFarm(metaclass=abc.ABCMeta):
+    """Abstract class representing a build farm managing multiple build hosts (request, wait, release, etc).
+
+    Attributes:
+        build_hosts: List of build hosts used for builds.
+        args: Set of args/options associated with the build farm.
+    """
+    build_hosts: List[BuildHost]
+    args: Dict[str, Any]
+
+    def __init__(self, args: Dict[str, Any]) -> None:
+        """
+        Args:
+            args: Args (i.e. options) passed to the build farm.
+        """
+        self.args = args
+        self.build_hosts = []
+
+    @staticmethod
+    @abc.abstractmethod
+    def NAME() -> str:
+        """Human-readable name (used as the "build-farm-type" in the YAML).
+
+        Returns:
+            Human-readable name.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def parse_args(self) -> None:
+        """Parse default build farm arguments."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def request_build_host(self, build_config: BuildConfig) -> None:
+        """Request build host to use for build config.
+
+        Args:
+            build_config: Build config to request build host for.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def wait_on_build_host_initialization(self, build_config: BuildConfig) -> None:
+        """Ensure build host is launched and ready to be used.
+
+        Args:
+            build_config: Build config used to find build host that must ready.
+        """
+        raise NotImplementedError
+
+    def get_build_host(self, build_config: BuildConfig) -> BuildHost:
+        """Get build host associated with the build config.
+
+        Args:
+            build_config: Build config used to find the build host.
+
+        Returns:
+            Build host associated with the build config.
+        """
+        for build_host in self.build_hosts:
+            if build_host.build_config == build_config:
+                return build_host
+
+        raise Exception(f"Unable to find build host for {build_config.name}")
+
+    def get_build_host_ip(self, build_config: BuildConfig) -> str:
+        """Get IP address associated with this dispatched build host.
+
+        Args:
+            build_config: Build config to find build host for.
+
+        Returns:
+            IP address for the specific build host.
+        """
+        return self.get_build_host(build_config).ip_address
+
+    @abc.abstractmethod
+    def release_build_host(self, build_config: BuildConfig) -> None:
+        """Release the build host.
```

Comment:
I think it would be good to explain what this means in greater detail here. 
For cloud-based deployments, it may result in the termination of instanes to save cost. In other cases it may enable reusing the host as part of another buildfarm.

Reply:
This is left up to the buildfarm implementation of the function to describe more. I kept this intentionally vague so that people would look at the other comment for the specifics.